Kernel runtime error checking

nashif opened the pull request:
Introduce runtime error checking in the kernel where it make sense and remove duplication of asserts and oopses that were used to catch such errors. Propagate errors where needed. 

Ideally we want to use this to check for parameters and permissions and return to the caller to allow recovery, which was not possible in the current approach: when asserts are enabled (not default) we would just crash on every error, even if recoverable or we will get undefined behavior if asserts were not enabled.

Situations which should never happen and disallowed by design, such as running something in an ISR when it should not, remain as asserts where applicable.

-----------------

(Edit by @aescolar) Discussed in #17105

nashif committed:
kernel: mutex: add error checking

k_mutex_unlock will now perform error checking and return on failures.

If the current thread does not own the mutex, we will now return -EPERM.
In the unlikely situation where we own a lock and the lock count is
zero, we assert. This is considered an undefined bahviour and should not
happen.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: semaphore: k_sem_init error checking

Check for errors at runtime and stop depending on ASSERTs.
This changes the API for
- k_sem_init

k_sem_init now returns -EINVAL on invalid data.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: pipe: runtime error checking

Add runtime error checking to k_pipe_cleanup and k_pipe_get and remove
asserts.
Adapted test which was expecting a fault to handle errors instead.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: msgq: error handling

Add runtime error handling for k_msgq_cleanup. We return 0 on success
now and -EAGAIN when cleanup is not possible.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: work_q: error handling

When trying to cancel a NULL work queue return -EAGAIN.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: stack: error handling

Add runtime error checking for both k_stack_push and k_stack_cleanup.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: mem_slab: error handling

Add runtime error checking for k_mem_slab_init and replace asserts with
returning error codes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: queue: runtime error handling

Runtime error handling for k_queue_append_list and k_queue_merge_slist.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ceolin commented on include/kernel.h:
```diff
@@ -3030,10 +3030,12 @@ __syscall int k_mutex_lock(struct k_mutex *mutex, s32_t timeout);
- * @return N/A
+ * @retval 0 Mutex unlocked.
+ * @retval -EPERM The current thread does not own the mutex.
+ * @retval -EINVAL Undefined behavior.
```
EINVAL is for invalid parameter not undefined behavior

ceolin replied:
btw, return EINVAL seems fine, can't we just say that mutex is not locked

nashif replied:
I am going to follow posix here and return EPERM in both cases

pfl commented on kernel/poll.c:
```diff
@@ -35,13 +35,17 @@
-	__ASSERT(obj != NULL, "must provide an object\n");
+	/* only NOTIFY_ONLY mode is supported */
+	if (mode != K_POLL_MODE_NOTIFY_ONLY) {
+		return -ENOTSUP;
```
-ENOTSUP is not documented in the .h file?

pfl commented on kernel/queue.c:
```diff
@@ -255,11 +258,18 @@ void k_queue_append_list(struct k_queue *queue, void *head, void *tail)
+	/* list must not be empty */
+	if (sys_slist_is_empty(list)) {
+		return -EINVAL;
+	}
```
Nitipick: if the list is empty, the function could just return here?

andrewboie commented on kernel/mem_slab.c:
```diff
@@ -28,15 +28,16 @@ struct k_mem_slab *_trace_list_k_mem_slab;
  * @return N/A
  */
-static void create_free_list(struct k_mem_slab *slab)
+static int create_free_list(struct k_mem_slab *slab)
```
not all callers of this function are checking the return value.

andrewboie commented on kernel/msg_q.c:
```diff
@@ -94,14 +94,17 @@ Z_SYSCALL_HANDLER(k_msgq_alloc_init, q, msg_size, max_msgs)
 {
-	__ASSERT_NO_MSG(z_waitq_head(&q->wait_q) == NULL);
+	if (z_waitq_head(&q->wait_q) != NULL) {
+		return -EAGAIN;
```
maybe -EBUSY would be more fitting?

andrewboie commented on kernel/mutex.c:
```diff
@@ -200,18 +182,32 @@ Z_SYSCALL_HANDLER(k_mutex_lock, mutex, timeout)
+	 * below.
+	 */
+	if (mutex->lock_count == 0U) {
+		return -EPERM;
```
Please use the same error codes as sys_mutex(), for consistency.
This should be -EINVAL

tyrantlizard77 reviewed (changes requested):
You can get rid of some of the checks in sys_mutex code since the base implementation is now doing it.

dcpleung commented on kernel/mutex.c:
```diff
@@ -203,12 +203,24 @@ static inline int z_vrfy_k_mutex_lock(struct k_mutex *mutex, s32_t timeout)
 
-	__ASSERT(mutex->lock_count > 0U, "");
-	__ASSERT(mutex->owner == _current, "");
+	/*
```
I think permission check should be first so it won't leak information about a non-owned mutex.

ceolin commented on kernel/pipes.c:
```diff
@@ -173,16 +173,19 @@ static inline int z_vrfy_k_pipe_alloc_init(struct k_pipe *pipe, size_t size)
-	__ASSERT_NO_MSG(!z_waitq_head(&pipe->wait_q.readers));
-	__ASSERT_NO_MSG(!z_waitq_head(&pipe->wait_q.writers));
+	if (z_waitq_head(&pipe->wait_q.readers) ||
+			z_waitq_head(&pipe->wait_q.writers)) {
```
can you make it boolean operations ? e.g `z_waitq_head(&pipe->wait_q.readers) != NULL`

ceolin commented on include/kernel.h:
```diff
@@ -1543,9 +1543,10 @@ typedef void (*k_timer_stop_t)(struct k_timer *timer);
  * @param stop_fn   Function to invoke if the timer is stopped while running.
  *
- * @return N/A
+ * @retval 0 on success
```
why change it if this function always return 0 ? There is nothing to check here.

ceolin commented on kernel/mutex.c:
```diff
@@ -130,10 +124,6 @@ int z_impl_k_mutex_lock(struct k_mutex *mutex, s32_t timeout)
-		K_DEBUG("%p took mutex %p, count: %d, orig prio: %d\n",
-			_current, mutex, mutex->lock_count,
-			mutex->owner_orig_prio);
-
```
+1 for removing K_DEBUG, but wouldn't be good keep this info ? we could call printk directly under `ifdef CONFIG_DEBUG` guard.

ceolin commented on include/kernel.h:
```diff
@@ -3448,11 +3448,19 @@ void k_msgq_init(struct k_msgq *q, char *buffer, size_t msg_size,
+ *
+ * @retval 0 on success
+ * @retval -EAGAIN Queue not empty
+ */
```
s/EAGAIN/EBUSY/

ceolin replied:
or change the return in the implementation :) but EBUSY seems better

ceolin replied:
though `k_stack_cleanup` is returning EAGAIN. Just need to be consistent :)

aescolar reviewed (dismissed):
@nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
* smaller embedded systems which will be too constrained to do all error checking in the app
* current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)

On https://github.com/zephyrproject-rtos/zephyr/issues/17105#issuecomment-508694903 I elaborated a bit more on the problem and on the kind of solution I really think would be best.
But if that is too cumbersome to implement, I would at least allow to kconfig if there is asserts or runtime checks, instead of just leaving only one option.

nashif commented:
> @nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
> 
>     * smaller embedded systems which will be too constrained to do all error checking in the app
> 
>     * current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)

@aescolar there is no intention to go and remove every assert we have and replace it with runtime error-checking. However, we do need to go through the usages of asserts and create some guidelines, because asserts in general are being misused and we are left in some cases with no runtime checking where it really matters, because someone decided it is easier to just use an ASSERT or because it has always been done this way, etc.

Regarding footprint in general we need to be talking in numbers here and not talk in abstracts ("smaller embedded systems...."), i.e. set targets, have budgets and put everyone and every subsystem under the same rules and guidelines and scrutiny. I am sure there is always more to be saved in terms of footprint above the kernel.

The kernel is not the only place where footprint should be saved, it is however the most importat area where runtime error checking is needed due to the fact that it can be the source for many issues showing up all the way up the stack if not handled correctly, both from quality, safety and security perspective.

ceolin commented:
> > @nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
> > ```
> > * smaller embedded systems which will be too constrained to do all error checking in the app
> > 
> > * current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)
> > ```
> 
> @aescolar there is no intention to go and remove every assert we have and replace it with runtime error-checking. However, we do need to go through the usages of asserts and create some guidelines, because asserts in general are being misused and we are left in some cases with no runtime checking where it really matters, because someone decided it is easier to just use an ASSERT or because it has always been done this way, etc.
> 
> Regarding footprint in general we need to be talking in numbers here and not talk in abstracts ("smaller embedded systems...."), i.e. set targets, have budgets and put everyone and every subsystem under the same rules and guidelines and scrutiny. I am sure there is always more to be saved in terms of footprint above the kernel.
> 
> The kernel is not the only place where footprint should be saved, it is however the most importat area where runtime error checking is needed due to the fact that it can be the source for many issues showing up all the way up the stack if not handled correctly, both from quality, safety and security perspective.

agreed ! Adding two more cents, don't have this kind of error checking within the kernel makes impossible do it properly on up layers. For instance, syscalls can not return proper errors if we don't have it done in inner functions.

aescolar commented:
@nashif : Ok, so given that this is WIP, and I cannot read your mind to know what is it that you plan to do,
instead of me trying to guess the extent of what you want to do, and probably talking past each other, can you describe what is it you are proposing?

My concerns and arguments I described on #17105 (comment). The main argument is: _depending on what you are trying to build, your needs can be very different_.

andrewboie commented:
> smaller embedded systems which will be too constrained to do all error checking in the app

Can we be data-driven here, and show some numbers on the footprint increase associated with this, before we sound the alarm?

aescolar commented:
@andrewboie @nashif 
Guys, it seems to me you are fixated thinking this is an "either or".
Doesn't doing something like this below cover everybody's needs?
(Please ignore naming)
```diff
diff --git a/kernel/sem.c b/kernel/sem.c
index f1fd430e64..c324bc315b 100644
--- a/kernel/sem.c
+++ b/kernel/sem.c
@@ -59,13 +59,27 @@ SYS_INIT(init_sem_module, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
 
 #endif /* CONFIG_OBJECT_TRACING */
 
+
+#if defined(ASSERT_ON_ERRORS_IN_KERNEL)
+       #define CHECK(expr) \
+                __ASSERT_NO_MSG(!(expr));   \
+                if (0)
+#elif defined(RETURN_ERRORS_IN_KERNEL)
+       #define CHECK(expr) \
+                if (expr)
+#elif defined(NO_KERNEL_RUNTIME_CHECKS)
+       #define CHECK(...) \
+                if (0)
+#endif
+
+
 int z_impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
                      unsigned int limit)
 {
        /*
         * Limit cannot be zero and count cannot be greater than limit
         */
-       if (limit == 0U || initial_count > limit) {
+       CHECK(limit == 0U || initial_count > limit) {
                return -EINVAL;
        }
```

andrewboie commented:
> Doesn't doing something like this below cover everybody's needs? (Please ignore naming)

Being specific really helps. This approach looks good to me. We'll need to ensure that if USERSPACE=y then the RETURN_ERRORS_IN_KERNEL option is selected, since the syscall verifiers will rely on them.

tyrantlizard77 reviewed (changes requested):
At this time I concur with @aescolar now that he has proposed an alternative to doing this which gives greater control on how much footprint to use for error handling.

nashif committed:
base: add error checking macros

Define there options for runtime error handling:
- assert on all errors (ASSERT_ON_ERRORS)
- no runtime checks (no asserts, no runtime error handling)
  (NO_RUNTIME_CHECKS)
- full runtime error handling (the default) (RUNTIME_ERROR_CHECKS)

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
using a CHECK macro now.

galak commented on Kconfig.zephyr:
```diff
@@ -248,6 +248,30 @@ config COMPILER_OPT
 endmenu
 
+
+
```
nit, extra whitespace/newline.

galak commented on include/kernel.h:
```diff
@@ -3586,11 +3598,19 @@ void k_msgq_init(struct k_msgq *q, char *buffer, size_t msg_size,
  * @req K-MSGQ-002
  */
-__syscall int k_msgq_alloc_init(struct k_msgq *q, size_t msg_size,
+__syscall int k_msgq_alloc_init(struct k_msgq *msgq, size_t msg_size,
```
can you pull the arg renaming into a different PR/commit?

galak commented on Kconfig.zephyr:
```diff
@@ -248,6 +248,30 @@ config COMPILER_OPT
+	  Do not do any runtime checks or asserts.
+
+config RUNTIME_ERROR_CHECKS
+	bool "No runtime checks"
```
the 'no runtimes checks' seems wrong here.

galak commented on include/kernel.h:
```diff
@@ -3610,7 +3630,7 @@ void k_msgq_cleanup(struct k_msgq *q);
  * @req K-MSGQ-002
  */
-__syscall int k_msgq_put(struct k_msgq *q, void *data, s32_t timeout);
+__syscall int k_msgq_put(struct k_msgq *msgq, void *data, s32_t timeout);
```
similar comment about arg renaming, just makes this a bit more difficult to review.

galak reviewed (dismissed):
Minor issue in the Kconfig that needs cleaning up.

aescolar commented on Kconfig.zephyr:
```diff
@@ -248,6 +248,27 @@ config COMPILER_OPT
+config RUNTIME_ERROR_CHECKS
+	bool "Enable runtime error checks"
+	help
+	  Do runtime checks, this is the default and the only option used
```
Nit: Maybe better rephrase "and the only option used" => "and the typical option used"
I guess we will need sometimes to disable input checks for some reasons (e.g. sample almost too big for microbit)

nashif replied:
This is going to be the only option enabled during testing, there is no way around it. Enable any of the other options and things will collapse now that we are going to do negative testing, the whole purpose of this action.

aescolar replied:
If this is indeed the only option used during testing, things will probably be broken quite soon for the other options. I would make sense to do most of the kernel testing with this one. But still, I would consider it reasonable to have a couple of tests, if only tests with of samples or higher functionality, which would use the other options.

aescolar commented on Kconfig.zephyr:
```diff
@@ -248,6 +248,27 @@ config COMPILER_OPT
+config NO_RUNTIME_CHECKS
+	bool "No runtime error checks"
+	help
+	  Do not do any runtime checks or asserts.
```
Nit: How it is phrased now, it seems to imply this option disables all asserts and error checking everywhere.

Would it be better to mention in the prompt above "Error checking behavior for CHECK macro" (and remove "covered with the CHECK macro" from the first option).
OR
repeat "covered with the CHECK macro" in the 2 others?

nashif replied:
sure.

aescolar reviewed (commented):
Thanks Anas, this should cover both needs nicely.
Just two minor comments

nashif committed:
tests: test k_sem_init()

Add tests for k_sem_init.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: semaphore: optimize code

Remove static helper functions used only once and integrate them into
calling functions.
In k_sem_take, return at the end.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

andrewboie commented on include/sys/check.h:
```diff
@@ -0,0 +1,26 @@
+	__ASSERT_NO_MSG(!(expr));   \
+	if (0)
+#elif defined(CONFIG_NO_RUNTIME_CHECKS)
+#define CHECK(...) \
```
Are there any cases where completely dropping the arguments could generate a compiler warning about "set but not used" or something like that?

andrewboie commented:
Any particular reason why the assertions in kernel/mem_domain.c were not updated with checks?

One final thought: what do we do about k_thread_create()?

nashif commented:
> One final thought: what do we do about k_thread_create()?

what about it? there is one assert in there that should be left as such (creating thread from ISR). Are you referring to the load of Z_OOPSes in z_vrfy_k_thread_create?

nashif commented:
> Any particular reason why the assertions in kernel/mem_domain.c were not updated with checks?

No reason, first I wanted us to agree this approach is the one to follow before moving too fast, there is also poll.c which also needs conversion.

andrewboie commented:
> what about it? there is one assert in there that should be left as such (creating thread from ISR). Are you referring to the load of Z_OOPSes in z_vrfy_k_thread_create?

Yeah -- I'd rather propagate EINVAL somehow.
But since it returns a pointer...maybe it's enough to return NULL if a check fails.

tyrantlizard77 reviewed (approved):
Need to get CI passing, and please verify that choices won't inadvertently cause compiler warnings, otherwise this approach LGTM.

nashif commented:
> For some weird reason this very old review is still here. It does not apply anymore. The later review superseded it

so, you are not approving?

aescolar reviewed (dismissed):
> so, you are not approving?

I'm not the kernel maintainer, so it is for me to give a +2.
My concerns have been addressed. So this is just a +0.

ioannisg commented on kernel/mem_slab.c:
```diff
@@ -57,31 +60,45 @@ static void create_free_list(struct k_mem_slab *slab)
-	create_free_list(slab);
+	rc = create_free_list(slab);
+	if (rc < 0) {
+		goto out;
```
early return here, instead?

ioannisg reviewed (commented):
some minor comments from my side

andrewboie commented:
> My concerns have been addressed. So this is just a +0.

@aescolar this isn't helpful.
You obviously have strong opinions about this topic, you are allowed to give a +1

aescolar commented:
> you are allowed to give a +1

@andrewboie Thanks. I understand I'm allowed to approve the PR. But the thing is that I do not support the overall direction things are going, but I will not oppose it either while the inconvenience for my company or others trying to target constrained platforms are not too great.

ioannisg commented:
> > you are allowed to give a +1
> 
> @andrewboie Thanks. I understand I'm allowed to approve the PR. But the thing is that I do not support the overall direction things are going, but I will not oppose it either while the inconvenience for my company or others trying to target constrained platforms are not too great.

@aescolar @nashif as far I understood, it is possible for this patch to have no overhead, if the introduced Kconfig switches are set properly, i.e. disable both ASSERT and NO_RUNTIME error checking, am I correct?

aescolar commented:
@ioannisg My concern is two fold:
* My approval of this PR may be miss-represented as support of further changes. And it is not clear to me what will happen after.
* Already @nashif is implying that testing with anything else than RUNTIME_ERROR_CHECKS will not be done.

andrewboie commented:
>  But the thing is that I do not support the overall direction things are going

What? Be specific

aescolar commented:
> What? Be specific

Disregard for platform constrained targets.
Over-prioritization of some assumed to be requirements of functional safety at the expense of any other target for the project.

nashif commented:
> Over-prioritization of some assumed to be requirements of functional safety at the expense of any other target for the project.

This has nothing to do with functional safety. The fact that we are not able to test basic APIs because we just crash is bad for EVERYONE.

andrewboie commented:
> Disregard for platform constrained targets.

@aescolar let's explore this.
How does this PR disregard platform constrained targets?
Be specific.

aescolar commented:
> @aescolar let's explore this.
> How does this PR disregard platform constrained targets?

The fact that I removed my -1 would be an indication that it does not anymore. Wouldn't it?
